packages/transformers/postcss/src/loadConfig.js

```diff
@@ -0,0 +1,149 @@
+// @flow
+import type {Config, PluginOptions} from '@parcel/types';
+import type {PluginLogger} from '@parcel/logger';
+import path from 'path';
+
+import loadExternalPlugins from './loadPlugins';
+
+const MODULE_BY_NAME_RE = /\.module\./;
+
+async function configHydrator(
+  configFile: any,
+  config: Config,
+  options: PluginOptions,
+) {
+  // Use a basic, modules-only PostCSS config if the file opts in by a name
+  // like foo.module.css
+  if (configFile == null && config.searchPath.match(MODULE_BY_NAME_RE)) {
+    configFile = {
+      plugins: {
+        'postcss-modules': {},
+      },
+    };
+  }
+
+  if (configFile == null) {
+    return;
+  }
+
+  // Load the custom config...
+  let modules;
+  let configPlugins = configFile.plugins;
+  if (
+    configPlugins != null &&
+    typeof configPlugins === 'object' &&
+    configPlugins['postcss-modules'] != null
+  ) {
+    modules = configPlugins['postcss-modules'];
+    // $FlowFixMe
+    delete configPlugins['postcss-modules'];
+  }
+
+  if (!modules && configFile.modules) {
+    modules = {};
+  }
+
+  let plugins = await loadExternalPlugins(
+    configPlugins,
+    config.searchPath,
+    options,
+  );
+
+  return config.setResult({
+    raw: configFile,
+    hydrated: {
+      plugins,
+      from: config.searchPath,
+      to: config.searchPath,
+      modules,
+    },
+  });
+}
+
+export async function load({
+  config,
+  options,
+  logger,
+}: {|
+  config: Config,
+  options: PluginOptions,
+  logger: PluginLogger,
+|}) {
+  let configFile: any = await config.getConfig(
+    ['.postcssrc', '.postcssrc.json', 'postcss.config.js', '.postcssrc.js'],
+    {packageKey: 'postcss'},
+  );
+
+  let configPath = config.resolvedPath;
+  if (configPath) {
+    if (path.extname(configPath) === '.js') {
+      logger.warn({
+        message:
+          'WARNING: Using a JavaScript PostCSS config file means losing out on caching features of Parcel. Use a .postcssrc(.json) file whenever possible.',
+      });
+
+      config.shouldInvalidateOnStartup();
+    }
+
+    if (configFile) {
+      if (typeof configFile !== 'object') {
+        throw new Error('PostCSS config should be an object.');
+      }
+
+      if (
+        configFile.plugins == null ||
+        typeof configFile.plugins !== 'object' ||
+        Object.keys(configFile.plugins) === 0
+      ) {
+        throw new Error('PostCSS config must have plugins');
+      }
+
+      let configFilePlugins = Array.isArray(configFile.plugins)
+        ? configFile.plugins
+        : Object.keys(configFile.plugins);
+      for (let p of configFilePlugins) {
+        // JavaScript configs can use an array of functions... ugh
+        if (typeof p === 'function') {
```

Comment:
Not entirely sure if there is a way to tell the config object something is not serialisable?

cc @devongovett @padmaia

Reply:
You mean something generic that wouldn't require knowledge of the config format? That would be interesting.

Reply:
I mean tell the config that there's no way to serialise this as it contains functions.
Unless we'd eval or new Function them afterwards although that sounds like a terrible idea

Reply:
Oh I see. Yeah right now it wouldn't matter because anytime we rerun a transformation we reloading all of it's configs. I think we were, but there was a bit of a regression when refactoring everything to run in the worker thread. Pretty sure the `shouldReload` helped with that now that I'm thinking about it. Maybe we should keep the function and use it when we start using cached configs for requests that weren't invalidated again. (Maybe that's why I kept it around)

I was thinking of a common util that would walk over object properties and array values recursively or something to determine if it could not be serialized for some reason, but I'm not sure how useful that would actually be.

Reply:
The util seems useful to me, would be nice if serialisation would just be wrapped in a `try {} catch {}` or something and on error (aka cannot be serialised) calls the `shouldReload()` automatically.

It would also significantly simplify this part of the code and be a lot easier to setup on even complexer config types.